_posts/2020-11-03-torch-tabular/torch_tabular.Rmd

```diff
@@ -0,0 +1,445 @@
+
+In a nutshell, what we'll see is
+
+-   How to create a custom *dataset*, tailored to the specific data you have.
+
+-   How to handle a mix of numeric and categorical data.
+
+-   How to extract the continuous-space representations from the embedding modules.
+
+## Dataset
+
+The dataset, [Mushrooms](https://archive.ics.uci.edu/ml/datasets/Mushroom), was chosen for its abundance of categorical columns. It is an unusual dataset to use in DL: It was designed for machine learning models to infer logical rules, as in: IF *a* AND NOT *b* OR *c* [...], then it's an *x*.
+
+Concretely, mushrooms are classified into two groups: edible and non-edible. The dataset description lists five possible rules with their resulting accuracies. While the least we want to go into here is the hotly debated topic of whether DL is suited to, or how it could be made more suited to, rule learning, we'll allow ourselves some curiosity and check out what happens if we successively remove all columns used to construct those five rules.
+
+Oh, and before you start copy-pasting: Here is the example in a [Google Colaboratory notebook](https://colab.research.google.com/drive/1bfyoD13YaLPLVcQVOLimv6vq9y4CKMM2?usp=sharing).
+
+```{r}
+library(torch)
+library(purrr)
+library(readr)
+library(dplyr)
+library(ggplot2)
+
+download.file(
+  "https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.data",
+  destfile = "agaricus-lepiota.data"
+)
+
+mushroom_data <- read_csv(
+  "agaricus-lepiota.data",
+  col_names = c(
+    "poisonous",
+    "cap-shape",
+    "cap-surface",
+    "cap-color",
+    "bruises",
+    "odor",
+    "gill-attachment",
+    "gill-spacing",
+    "gill-size",
+    "gill-color",
+    "stalk-shape",
+    "stalk-root",
+    "stalk-surface-above-ring",
+    "stalk-surface-below-ring",
+    "stalk-color-above-ring",
+    "stalk-color-below-ring",
+    "veil-type",
+    "veil-color",
+    "ring-type",
+    "ring-number",
+    "spore-print-color",
+    "population",
+    "habitat"
+  ),
+  col_types = rep("c", 23) %>% paste(collapse = "")
+) %>%
+  # can as well remove because there's just 1 unique value
+  select(-`veil-type`)
+```
+
+In `torch`, `dataset()` creates an R6 class. Like with most R6 classes, there will usually be a need for an `initialize()` method. Below, we use `initialize()` to preprocess the data and store it in convenient pieces. More on that in a minute. Prior to that, please note the two other methods a `dataset` has to implement:
+
+-   `.getitem(i)` . This is the whole purpose of a `dataset`: Retrieve and return the observation located at some index it is asked for. Which index? That's to be decided by the caller, a `dataloader`. During training, usually we want to permute the order in which observations are used, while not caring about order in case of validation or test data.
```

Comment:
Just double checking that when you're referencing `dataset` and `dataloader` you're talking about them as nouns (and not functions, in which case you'd follow them with `()`).

Reply:
yeah, great question :-) In Python this would be less awkward because these would clearly and visibly be classes as opposed to functions